pkg/virt-controller/watch/migration.go

```diff
@@ -1029,3 +1047,41 @@ func (c *MigrationController) findRunningMigrations() ([]*virtv1.VirtualMachineI
 	}
 	return runningMigrations, nil
 }
+
+func (c *MigrationController) getNodeForVMI(vmi *virtv1.VirtualMachineInstance) (*k8sv1.Node, error) {
+	obj, exists, err := c.nodeInformer.GetStore().GetByKey(vmi.Status.NodeName)
+
+	if err != nil {
+		return nil, fmt.Errorf("cannot get nodes to migrate VMI with host-model CPU. error: %v", err)
+	} else if !exists {
+		return nil, fmt.Errorf("node \"%s\" assosiated with vmi \"%s\" does not exist", vmi.Status.NodeName, vmi.Name)
+	}
+
+	node := obj.(*k8sv1.Node)
+	return node, nil
+}
+
+func prepareNodeSelectorForHostCpuModel(node *k8sv1.Node, pod *k8sv1.Pod) error {
+	var hostCpuModel string
+	var labelValue string
+
+	for key, value := range node.Labels {
+		if strings.HasPrefix(key, virtv1.HostModelCPULabel) {
+			hostCpuModel = strings.TrimPrefix(key, virtv1.HostModelCPULabel)
+			labelValue = value
+		}
+
+		if strings.HasPrefix(key, virtv1.HostModelRequiredFeaturesLabel) {
+			requiredFeature := strings.TrimPrefix(key, virtv1.HostModelRequiredFeaturesLabel)
+			pod.Spec.NodeSelector[virtv1.CPUFeatureLabel+requiredFeature] = value
+		}
+	}
+
+	if hostCpuModel == "" {
+		return fmt.Errorf("node does not contain labal \"%s\" with information about host cpu model", virtv1.HostModelCPULabel)
+	}
```

Comment:
I'm hesitant for us to block the migration if hostCpuModel isn't found.

the node labeler isn't always present. for example on arm64 and ppc64 architectures it won't run today. I think it is disabled for emulation mode as well.

does it make sense for this to be a best effort approach instead? If the host cpu model and required features are found, use them for scheduling, and if they aren't then skip?

Reply:
Hey David!
I'm afraid that if we would do that we will not fix [the bug](https://bugzilla.redhat.com/show_bug.cgi?id=1760028) that this PR aims to fix.

If the VM uses host-model CPU mode and we have no idea which model it is nor which node supports this model then we cannot know which node is valid for scheduling. Today we allow the migration but if the CPU model is not supported on the node the above bug will appear.

As I see it, if we want to always avoid this bug, we have no other option than to fail the migration if we cannot know the actual CPU model.

However, it is reasonable IMO to write in user-guide that in order for VMIs with host-model CPU to migrate correctly node-labeller has to be do its work, and if it doesn't, then we would migrate anyways and hope that we're lucky and the VM lands on a node which supports the right CPU model.

Is that what we want to do? What do you think?

Reply:
I see your point.
Today, when a VMI with host-model CPU migrates to a node that doesn't support this model the error is the following (as appears in [this bug](https://bugzilla.redhat.com/show_bug.cgi?id=1760028)):
`
{"component":"virt-launcher","kind":"","level":"error","msg":"Live migration failed","name":"vm-fedora","namespace":"default","pos":"manager.go:473","reason":"virError(Code=9, Domain=31, Message='operation failed: guest CPU doesn't match specification: missing features: md-clear')","timestamp":"2019-10-09T15:43:59.828172Z","uid":"2ce9e37d-eaab-11e9-ba41-fa163e35f769"}
`

This means we ask for an invalid operation and depend on libvirt to validate us, which is a bit of a concern. But maybe a warning is enough in case we cannot validate on our own.

P.S. I think it also has something to do with the feature gate discussion on which I've commented below. If we have a feature gate a user that expects migrations of such VMIs is more likely to understand where such an error is coming from.